### ISSUE TITLE
IE 7 and 8 support

### ISSUE THREAD

**Issue by accursoft** [accursoft/caret#2]
*Original report by Anonymous.*

----------------------------------------

The plugin does not work in ie7 and 8 (I actually tested in IE 9 compatibility mode, but I think I will get the same result in real IE7-8)

I found another example of cursor manipulation(http://blog.vishalon.net/index.php/javascript-getting-and-setting-caret-position-in-textarea) and fixed you code according to it

Getting position

```
#!javascript

 //IE<9
 if (document.selection) {
	target.focus();
	var sel = document.selection.createRange();
	sel.moveStart('character', -target.value.length);
	return sel.text.length;
  }
```
Setting position 

```
#!javascript

//IE<9
else if (document.body.createTextRange) {
	var range = target.createTextRange();
	range.collapse(true);
	range.moveEnd('character', pos);
	range.moveStart('character', pos);
	range.select();	
}
```

With these fixes it work fine in old IEs, please apply them to your code

**Comment by accursoft** [comment]
I've tested this on both IE6 and IE8. If you find it to be broken on an actual browser instance (as opposed to compatibility mode), please reopen the issue with a concrete example of where it fails.

**Comment by accursoft** [comment]
*Original comment by Simon Steinberger (Bitbucket: [SimonStone](https://bitbucket.org/SimonStone), GitHub: [SimonStone](https://github.com/SimonStone)).*

----------------------------------------

There is definitely still a bug with IE7 and IE8. It works for me by changing these lines:

```
#!javascript

var range = document.body.createTextRange();
range.moveToElementText(target);
range.moveStart('character', pos);
range.collapse(true);
range.select();

```

Into:


```
#!javascript

var range = target.createTextRange();
range.move('character', pos);
range.select();

```

Otherwise great plugin - thanks!

**Comment by accursoft** [comment]
It works for me in both IE7 and IE8. Can you be more specific about what the problem is? Can you reproduce it with the included test page (`test.html`)?

**Comment by accursoft** [comment]
*Original comment by Simon Steinberger (Bitbucket: [SimonStone](https://bitbucket.org/SimonStone), GitHub: [SimonStone](https://github.com/SimonStone)).*

----------------------------------------

Thanks for your quick reply. It would be great to get this working as intended. Phew! I've just tested a lot and debugged again.

"test.html" works nicely. **But** there's another scenario that fails. JSFiddle doesn't work with IE7-8. So I just post an extended "test.html" right here. Hope that's okay.

```
#!html

<!DOCTYPE html>
<html>
<head>
<script type="text/javascript" src="http://code.jquery.com/jquery.min.js"></script>
<script type="text/javascript" src="jquery.caret.js"></script>
<script type="text/javascript">
$(function() {
    var element = $('#text');
    var text = element.val();
    element.caret(2);
    var pos = element.caret();
    element.val(text.substring(0, pos) + 'b' + text.substring(pos));

    element = $('#editable');
    text = element.html();
    element.caret(1);
    pos = element.caret();
    element.html(text.substring(0, pos) + 'b' + text.substring(pos));

    $('#click_test').click(function(){
        var input = $(this).html('<input type="text" value="test text">').find('input');
        input.focus().caret(-1);
    });
});
</script>
</head>
<body>
<div id="click_test">Click me bug</div>
<textarea id="text">a

c</textarea>
<input type="text" id="pos">
<button onclick="$('#pos').val($('#text').caret())">read</button>
<button onclick="$('#text').caret($('#pos').val())">write</button>
<div id="editable" contenteditable="true">ac</div>
</body>
</html>
```

When clicking on the "Click me bug" DIV, the caret should be placed at the end of the input field. Yet, positioning doesn't work here - no matter which position. AFAICS, that's only a problem with input fields, not with textareas.

However, I just noticed, my above posted code doesn't completely fix this. The previous code works on textareas, my code works with input fields. So, this one appears to do the job well in all cases:

```
#!javascript

    //IE<9
    else if (document.body.createTextRange) {
      if (isContentEditable) {
        var range = document.body.createTextRange();
        range.moveToElementText(target);
        range.moveStart('character', pos);
        range.collapse(true);
        range.select();
      } else {
        var range = target.createTextRange();
        range.move('character', pos);
        range.select();
      }
    }

```
Maybe this can be shortened a bit ... al least it works :-) Looking forward to your feedback.

**Comment by accursoft** [comment]
fixed in f66cf605ca7d79768e18e84a46840147f294f310
